Makefile

```diff
@@ -135,10 +139,122 @@ proto-gen: install-proto-bin install-license-bin
+
+# Map generation rule for storage package maps (to avoid double build over each other
+# when generating map source files in parallel, run these sequentially)
+.PHONY: map-storage-gen
+map-storage-gen:
+	make map-storage-database-namespaces-gen
+	make map-storage-shard-gen
+
+# Map generation rule for storage/databaseNamespacesMap
+.PHONY: map-storage-database-namespaces-gen
+map-storage-database-namespaces-gen: install-m3x-repo
+	cd $(m3x_package_path) && make idhashmap-gen \
+		pkg=storage \
+		value_type=databaseNamespace \
+		out_dir=$(m3db_package_path)/storage \
+		rename_type_prefix=databaseNamespaces \
+		rename_constructor=newDatabaseNamespacesMap \
+		rename_constructor_options=databaseNamespacesMapOptions
+	# Rename both generated map and constructor files
+	mv -f $(m3db_package_path)/storage/map_gen.go $(m3db_package_path)/storage/namespace_map_gen.go
+	mv -f $(m3db_package_path)/storage/new_map_gen.go $(m3db_package_path)/storage/namespace_new_map_gen.go
+
+# Map generation rule for storage/shardMap
+.PHONY: map-storage-shard-gen
+map-storage-shard-gen: install-m3x-repo
+	cd $(m3x_package_path) && make idhashmap-gen \
+		pkg=storage \
+		value_type=listElement \
+		out_dir=$(m3db_package_path)/storage \
+		rename_type_prefix=shard \
+		rename_constructor=newShardMap \
+		rename_constructor_options=shardMapOptions
+	# Rename both generated map and constructor files
+	mv -f $(m3db_package_path)/storage/map_gen.go $(m3db_package_path)/storage/shard_map_gen.go
+	mv -f $(m3db_package_path)/storage/new_map_gen.go $(m3db_package_path)/storage/shard_new_map_gen.go
+
+# Map generation rule for storage/namespace/metadataMap
+.PHONY: map-storage-namespace-metadata-gen
+map-storage-namespace-metadata-gen: install-m3x-repo
+	cd $(m3x_package_path) && make idhashmap-gen \
+		pkg=namespace \
+		value_type=Metadata \
+		out_dir=$(m3db_package_path)/storage/namespace \
+		rename_type_prefix=metadata \
+		rename_constructor=newMetadataMap \
+		rename_constructor_options=metadataMapOptions
+	# Rename both generated map and constructor files
+	mv -f $(m3db_package_path)/storage/namespace/map_gen.go $(m3db_package_path)/storage/namespace/metadata_map_gen.go
+	mv -f $(m3db_package_path)/storage/namespace/new_map_gen.go $(m3db_package_path)/storage/namespace/metadata_new_map_gen.go
+
+# Map generation rule for storage/repair/Map
+.PHONY: map-storage-repair-gen
+map-storage-repair-gen: install-m3x-repo
+	cd $(m3x_package_path) && make idhashmap-gen \
+		pkg=repair \
+		value_type=ReplicaSeriesBlocksMetadata \
+		out_dir=$(m3db_package_path)/storage/repair
+
+# Map generation rule for all generated maps
+.PHONY: map-all-gen
+map-all-gen: map-client-received-blocks-gen map-storage-block-retriever-gen map-storage-bootstrap-result-gen map-storage-gen map-storage-namespace-metadata-gen map-storage-repair-gen
+
+# Tests that all currently generated maps match their contents if they were regenerated now
+.PHONY: test-map-all-gen
+test-map-all-gen: map-all-gen
```

Comment:
should we do this for all codegen? can open an issue to track if so

Reply:
Sure thing, opened issue: 
https://github.com/m3db/m3db/issues/533